test/tools/cli/IonProcess.java

```diff
@@ -0,0 +1,314 @@
+                        .writeOutput(ionWriterForOutput, ionWriterForErrorReport);
+                ionReader.stepIn();
+
+                //recursive call
+                processEvent(ionWriterForOutput, ionWriterForErrorReport, ionReader, curTable);
+
+                //write a Container_End event and step out
+                new Event(EventType.CONTAINER_END,curType,null,null,
+                        null,null,null, curDepth )
+                        .writeOutput(ionWriterForOutput, ionWriterForErrorReport);
+                ionReader.stepOut();
+            } else {
+                ionStreamToEvent(EventType.SCALAR,ionReader).writeOutput(ionWriterForOutput, ionWriterForErrorReport);
+            }
+        }
+    }
+
+    /**
+     * this function is only used for CONTAINER_START and SCALAR eventType since other types don't need initial data.
+     */
+    private static Event ionStreamToEvent(EventType eventType, IonReader ionReader) throws Exception {
+        if(ionReader.getType() == null) return null;
+
+        IonType ionType = ionReader.getType();
+        SymbolToken field_name = ionReader.getFieldNameSymbol();
+        SymbolToken[] annotations = ionReader.getTypeAnnotationSymbols();
+
+        String value_text = null;
+        byte[] value_binary = null;
+
+        if(eventType == EventType.SCALAR) {
+            //write Text
+            ByteArrayOutputStream textOut = new ByteArrayOutputStream();
+            IonWriter textWriter = IonTextWriterBuilder.standard().build(textOut);
+            textWriter.writeValue(ionReader);
+            textWriter.finish();
+            value_text = textOut.toString();
+            //write binary
+            ByteArrayOutputStream binaryOut = new ByteArrayOutputStream();
+            IonWriter binaryWriter = IonBinaryWriterBuilder.standard().build(binaryOut);
+            binaryWriter.writeValue(ionReader);
+            binaryWriter.finish();
+            value_binary = binaryOut.toByteArray();
+        }
+
+        ImportDescriptor[] imports = null;
+
+        int depth = ionReader.getDepth();
+
+        return new Event(eventType,ionType,field_name,annotations,value_text,value_binary,imports,depth);
+    }
+
+    private static boolean isSameSymbolTable(SymbolTable x, SymbolTable y){
+        if(x.isSystemTable() && y.isSystemTable()){
+            return (x.getVersion() == y.getVersion());
+        } else if (x.isSharedTable() && y.isSharedTable()) {
+            return (x.getName() == y.getName() & (x.getVersion()==y.getVersion()));
+        } else if (x.isLocalTable() && y.isLocalTable()){
+            SymbolTable[] x_table = x.getImportedTables();
+            SymbolTable[] y_table = x.getImportedTables();
+            if(x_table.length != y_table.length){
+                for(int i=0; i<x_table.length; i++){
+                    if(!isSameSymbolTable(x_table[i], y_table[i]))  return false;
+                }
+            }
+            return true;
```

Comment:
In addition to the imports, you also need to check the local symbols lists.

Reply:
Checked (IonProcess.java 244-255)

Questions: 
1. I tried some example on ion-c, it treats the symbol tables with different symbols as same as long as they have the same imports. 

2. Symbols' scope are before the next local symbol table or it will be accumulated every time pass a local symbol table?

Reply:
1. Right, you only really need to create a SYMBOL_TABLE event when the imports change. However in ion-c, there's a nice way of determining when this happens by registering a callback function with the reader, whereas in ion-java there is no similar mechanism (short of using a low-level reader and manually handling the symbol tables, which we'd like to avoid). So we're left needing to detect symbol table changes by comparing the symbol tables. I do think you may be able to simplify this by comparing just the imports of the two symbol tables and creating a SYMBOL_TABLE event when they are different.

2. A symbol table is in scope until either a) an Ion version marker is encountered, or b) a new non-append local symbol table is encountered. Note: appended local symbol tables add new symbols without resetting the previously-declared symbols. It is also not possible to append imports, so appended local symbol tables never need a SYMBOL_TABLE event.

Reply:
Thank you for the answer, I'll check symbols (IonProcess.java 244-255)

Reply:
Sorry, the conclusion I meant to to convey was this:

> I do think you may be able to simplify this by comparing just the imports of the two symbol tables and creating a SYMBOL_TABLE event when they are different.

This would allow you to simplify the current logic by not comparing local symbols. I recommend trying that and comparing it to ion-c's behavior. If they match, let's do it.

Reply:
Sure, I'll only check imports.

it performances same as ion-c now.